### PR TITLE
Fix `get` token recursion issue on TokenStorage for expired tokens

### PR DESCRIPTION
**Pull request by meslater1030**
This PR addresses issues #102 and #76 and it represents an alternative solution to PR #106.

When calling [`get`](https://github.com/OfficeDev/office-js-helpers/blob/master/src/authentication/token.manager.ts#L73) on the TokenStorage class the method checks to see whether the token has expired. If the token has expired it calls [`super.delete(tokenKey)`](https://github.com/OfficeDev/office-js-helpers/blob/master/src/authentication/token.manager.ts#L81) which lives on the parent Storage class. The Storage class delete method calls [`this.get(tokenKey)`](https://github.com/OfficeDev/office-js-helpers/blob/master/src/helpers/storage.ts#L125). Since `this` in this case is the TokenStorage class we get an infinite recursion.

I don't believe it's good practice to delete the expired token within the `get` method of the TokenStorage class since it's a side effect. If the token is expired then subsequent actions should be left to the caller. The authenticator appears to be the only caller of this method within the library and so I've updated that code to delete expired tokens if necessary. I'm aware this is a public method on TokenStorage and I'm open to suggestions about other possible alterations if this represents a breaking change for end users. However - it appears that the check for an expired token is the only additional functionality TokenStorage provides and since it does that in a way that fundamentally does not work I suspect this is a net gain.

I've added tests for the change.

I've also updated the typings for tapable as per the recommendations here: https://github.com/webpack/tapable/issues/53. A new install of this project otherwise throws errors.

Finally, my tests revealed that the reviver deserializer in the Storage class relies on an outdated REGEX matcher for ISO strings. I've updated that as well.

Thanks for your help and let me know if you have any questions.

### CODE REVIEW

**Comment by meslater1030** [comment]
@sumurthy or @Zlatkovsky can I get a review on this?

**Comment by Andruha4u** [comment]
Hello, currently using this one and will appreciate merge it to master, thank you ahead.

**Comment by super2ni** [comment]
Hey there,

Sorry for the noobish question but I am not very confortable  with Git. Seems like your PR corrects a bug I have on IE and I would like to install your PR has a NPM dependency.

What is the correct npm command to do that?

Thank you,
Denis

**Comment by techieanalyst** [comment]
I'm also having issue which has an error saying: 
> unable to delete 'Microsoft' from storage

When will this fix be merged into master?

File: src/authentication/authenticator.ts [hunk 1]
```diff
@@ -67,6 +67,10 @@ export class Authenticator {
       return Promise.resolve(token);
     }
 
+    if (hasTokenExpired) {
```
Comment (YannickRe) [inline src/authentication/authenticator.ts]: This fails on first load, if token doesn't already exist in the localStorage. Get returns `null` and `TokenStorage.hasExpired(token)` returns `true`. Better is to also check on actual existence of the key:
```suggestion
    if (token && hasTokenExpired) {
```

Comment (meslater1030) [reply src/authentication/authenticator.ts]: @YannickRe Thanks! I made that change. Let me know if there's anything else.

**Comment by Zlatkovsky** [comment]
Adding @LouMM , @sumurthy as the maintainers of this repo.

**Comment by roenylagan** [comment]
Hi guys! I'm just wondering if this fix will be pushed in earliest time possible. Thank you very much

**Comment by tomaszzmuda** [comment]
It would be nice if this will work soon :)

**Comment by Mathyaku** [comment]
I'm still having this problem. Anyone has a solution for that ?

**Comment by meslater1030** [comment]
Closing due to lack of response from maintainers.

**Pull request closed without merging**